SourceKitStressTester/Sources/StressTester/RequestDurationManager.swift

```diff
@@ -0,0 +1,101 @@
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2021 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Common
+import Foundation
+
+fileprivate extension Array where Element == Int {
+  /// Creates a logarithmic histogram. For each `key`, the `value` contains the
+  /// number of elements in this array that are smaller than `2 ^ key` but not
+  /// smaller than `2 ^ (key - 1)`.
+  var logHistogram: [Int: Int] {
+    var result: [Int: Int] = [:]
+    for value in self {
+      let log = Int(log2(Double(value))) + 1
+      result[log, default: 0] += 1
+    }
+    return result
+  }
+}
+
+/// Captures aggregated information about executing a certain request kind on a
+/// certain file.
+struct AggregatedRequestDurations: Codable {
+  static let empty = AggregatedRequestDurations(instructionCounts: [])
+
+  /// The log histogram created by the extension on `Array` above
+  var logHistogram: [Int: Int]
+
+  /// The total number of instructions executed by the requests of this type
+  /// `totalInstructions` / `requestsExecuted` gives the average number of
+  /// instructions per request.
+  var totalInstructions: Int
+
+  /// The total number of requests of a given kind executed for a given file
+  var requestsExecuted: Int
+
+  init(instructionCounts: [Int]) {
+    logHistogram = instructionCounts.logHistogram
+    totalInstructions = instructionCounts.reduce(0, { $0 + $1 })
+    requestsExecuted = instructionCounts.count
+  }
+
+  mutating func merge(other: AggregatedRequestDurations) {
+    self.logHistogram.merge(other.logHistogram, uniquingKeysWith: {
+      $0 + $1
+    })
+    self.totalInstructions += other.totalInstructions
+    self.requestsExecuted += other.requestsExecuted
+  }
+}
+
+/// Contains aggregated information for all request kinds of all files executed
+/// by the stress tester.
+fileprivate struct RequestDurations: Codable {
+  // FIXME: We should be using `RequestKind` as the inner key but that causes
+  // the inner dictionary to be serialized as an array (rdar://78099769)
+  /// Maps file paths to request kinds to aggregated request duration information
+  var files: [String: [String: AggregatedRequestDurations]]
+}
+
+/// Collects the durations that requests executed by the stress tester took and
+/// writes them to `jsonFile` where the durations are collected together with
+/// all other stress tester runs.
+/// We are aggregating the results early (just keeping track of the average for
+/// each request type and a logarithmic histogram because keeping track of all
+/// request durations would result in a JSON file that is too large to handle
+/// easily.
```

Comment:
Old comment, keep track of the number of instructions + total requests executed now.